Health API - Monitoring local disk health

gmarouli opened the pull request:
This PR introduces the local health monitoring functionality needed for #84811 . The monitor uses the `NodeService` to get the disk usage stats and determines the node's disk health.

When a change in the disk's is detected or when the health node changes, this class would be responsible to send the node's health to the health node. Currently this is simulated with a method that just logs the current health.

The monitor keeps the last reported health, this way, if something fails on the next check it will try to resend the new health state.

gmarouli committed:
Calculate free disk space

gmarouli committed:
Monitor disk health

gmarouli committed:
Small improvements

gmarouli commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into health-disk-monitor-health

gmarouli committed:
Move package

gmarouli committed:
Merge branch 'main' into health-disk-monitor-health

gmarouli commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into health-disk-monitor-health

gmarouli committed:
Implement send data when new health node detected

gmarouli committed:
Rephrase the javadoc of LocalHealthMonitor

gmarouli committed:
Ensure sending health is race free

gmarouli committed:
fix formatting

gmarouli committed:
Change export order

gmarouli committed:
polish

gmarouli committed:
fix formatting

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

andreidan commented on server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java:
```diff
@@ -0,0 +1,251 @@
+    // This method is synchronized to ensure that we keep track of the last reported health state to the health node.
+    // Note: this is for demonstration purposes, when we will be actually sending the information to the health node we might
+    // choose a different way to keep it thread safe.
+    private synchronized void sendHealth(IndividualNodeHealth currentHealth) {
```
No synchronisation is needed as the cluster applier thread is .. just one thread (cluster changed is called sequentially) 
I see visibility is ensured by the volatile variables - could you please document the reason why each volatile variable is needed to be volatile? ie. I think for some it's "safe publication of objects written to them" 

Alternatively, should we leave the "send of health" part to another PR to keep this one limited in scope?

gmarouli replied:
I will document the volatiles and leave this to the next PR, I still think there is a race but I can elaborate more at the next PR.

andreidan commented on server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java:
```diff
@@ -0,0 +1,251 @@
+                if (healthMetadataInitialized) {
+                    scheduleNextRunIfEnabled(TimeValue.timeValueMillis(1));
+                }
+            } else if (newHealthNodeSelected(event)) {
+                sendHealth(lastReportedHealth);
+            }
```
should this be in a separate PR that handles "sending of health data" ?

gmarouli replied:
I can put it in the next PR that I am working on, thinking about it kind of helped me set up the code but I can remove this and just log a line upon a change of the node's health. 

I will remove it.

andreidan commented on server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java:
```diff
@@ -0,0 +1,251 @@
+            if (healthMetadataInitialized == false) {
+                healthMetadataInitialized = HealthMetadata.getFromClusterState(event.state()) != null;
+                if (healthMetadataInitialized) {
+                    scheduleNextRunIfEnabled(TimeValue.timeValueMillis(1));
```
I think this should schedule based the monitoring based on the normal interval

gmarouli replied:
hm, you mean that if it gets rescheduled it should have the initial delay? 

I have to admit that when I see something executed at an interval I do not expect to have an initial delay. I expect that it's going to be executed asap and then wait for the interval before the next execution. For example let's say that the interval is 10 minutes. I would think it's wrong to wait for 10 minutes before running it for the first time.

What do you think?

andreidan commented on server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java:
```diff
@@ -0,0 +1,251 @@
+                return null;
+            }
+
+            FsInfo.Path leastAvailablePath = null;
+            for (FsInfo.Path info : nodeStats.getFs()) {
+                if (leastAvailablePath == null) {
+                    leastAvailablePath = info;
+                } else if (leastAvailablePath.getAvailable().getBytes() > info.getAvailable().getBytes()) {
+                    leastAvailablePath = info;
+                }
+            }
```
would we be interested in the path ES is configured to store data? ie. `path.data` ?

gmarouli replied:
Hm.... But that wouldn't work for non-data nodes, right?

But even for data nodes, the `DiskThresholdMonitor` also checks for the least available path and I think it would be nice to be consistent. 

What do you think?

andreidan commented on server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java:
```diff
@@ -0,0 +1,251 @@
+
+    private static final Logger logger = LogManager.getLogger(LocalHealthMonitor.class);
+
+    public static final Setting<TimeValue> INTERVAL_SETTING = Setting.timeSetting(
```
Should we document this new setting?

gmarouli replied:
I have a task in https://github.com/elastic/elasticsearch/issues/84811 to document all settings introduced as part of this and to add the troubleshooting page. I prefer to do all documentation at the end to avoid going live before we remove the feature flag.

andreidan reviewed (commented):
Thanks for working on this Mary

I left a first set of comments (and a suggestion to break this PR up)

gmarouli committed:
Merge branch 'main' into health-disk-monitor-health

gmarouli commented on server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java:
```diff
@@ -0,0 +1,251 @@
+
+    private void scheduleNextRunIfEnabled(TimeValue time) {
+        if (enabled && threadPool.scheduler().isShutdown() == false) {
+            scheduled = threadPool.schedule(this::monitorHealth, time, ThreadPool.Names.MANAGEMENT);
```
`scheduleUnlessShuttingDown` is not returning the `Scheduler.ScheduledCancellable` so we cannot cancel a scheduled run. However, I do think that id I rework the code a bit I can use it a make something even nicer than what we already have. So, I will give it a try.

gmarouli committed:
Remove sending data simulation

gmarouli committed:
Refactor the synchronization of monitoring

gmarouli commented:
Update:
- The sending is removed and left for the next PR.
- We use `scheduleUnlessShuttingDown` and used an `AtomicBoolean` to ensure the single execution of the monitoring. I like this approach more. I think this reads better and it doesn't do extra cancelling and scheduling. If there are multiple runs scheduled at the same time only one will run, the other ones will just do nothing. What do you think?

gmarouli commented:
Investigating the test failure. I can reproduce the failure locally too with this seed:
```


./gradlew ':server:test' --tests "org.elasticsearch.health.node.LocalHealthMonitorTests.testEnablingAndDisabling" -Dtests.seed=C307AA9DA1F7BA82 -Dtests.locale=es-CO -Dtests.timezone=Etc/Greenwich -Druntime.java=17 |  
-- | --



```

gmarouli committed:
No runs before metadata is initialized

gmarouli commented:
Test fixed. The issue was that we removed the `scheduled` field that was used both to cancel a run but also to signal not allow setting changes to schedule runs. Fixed by using `healthMetadataInitialized` for the same purpose.

gmarouli commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into health-disk-monitor-health

gmarouli commented:
@elasticmachine run elasticsearch-ci/packaging-tests-unix-sample

andreidan commented on server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java:
```diff
@@ -0,0 +1,233 @@
+        // Wait until every node in the cluster is upgraded to 8.4.0 or later
+        if (event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)) {
+            // Wait until the health metadata is available in the cluster state
+            if (healthMetadataInitialized == false) {
```
This flag is a little misleading as it waits for a bit more than just the metadata to be initialised (ie. nodes versions). Could we rename it to reflect that?

Also, if the service is enabled,disabled, and enabled again we might end up with 2 scheduled monitoring tracks? (since `setEnabled` doesn't check if we're already running?)

gmarouli replied:
About the scheduling two monitoring tasks. That's true, but then one will see that the `inProgress` flag is not false and it will just exit without doing anything or scheduling anything. I thought it was good trade-off compared to locking. What do you think?

andreidan replied:
Yes ! That's a good trade-off 🚀  Can you please document it?

gmarouli replied:
Done, let me know if it's clear.

gmarouli replied:
Also renamed `healthMetadataInitialized` :)

andreidan commented on server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java:
```diff
@@ -0,0 +1,233 @@
+                return null;
+            }
+
+            FsInfo.Path leastAvailablePath = null;
+            for (FsInfo.Path info : nodeStats.getFs()) {
+                if (leastAvailablePath == null) {
+                    leastAvailablePath = info;
+                } else if (leastAvailablePath.getAvailable().getBytes() > info.getAvailable().getBytes()) {
+                    leastAvailablePath = info;
+                }
+            }
```
Could this be a standalone method that encapsulates this logic? (and we can unit test)

gmarouli replied:
on it!

gmarouli replied:
See: https://github.com/elastic/elasticsearch/pull/88996

andreidan reviewed (commented):
Thanks for iterating on this Mary

Left a few more suggestions

gmarouli commented on server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java:
```diff
@@ -0,0 +1,233 @@
+
+    // Helper method that starts the monitoring without a delay.
+    private void scheduleNowIfEnabled() {
+        scheduleNextRunIfEnabled(new TimeValue(1));
```
Yes, good point!

gmarouli commented on server/src/main/java/org/elasticsearch/health/node/IndividualNodeHealth.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * The health of an individual node which consists by the health status of different resources, currently only disk space.
+ */
+record IndividualNodeHealth(Disk disk) {
```
Yes, for sure, we can always figure out a better set-up when we need to add another source.

gmarouli commented on server/src/main/java/org/elasticsearch/health/node/selection/HealthNode.java:
```diff
@@ -43,4 +44,13 @@ public static PersistentTasksCustomMetadata.PersistentTask<?> findTask(ClusterSt
     }
+
+    @Nullable
+    public static DiscoveryNode findHealthNode(ClusterState clusterState) {
```
Correct I will clean it up

gmarouli commented on server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java:
```diff
@@ -0,0 +1,233 @@
+            IndividualNodeHealth previousHealth = this.lastReportedHealth;
+            IndividualNodeHealth currentHealth = new IndividualNodeHealth(diskCheck.getHealth(healthMetadata, clusterState));
+            if (currentHealth.equals(previousHealth) == false) {
+                logger.info("Sending node health [{}] to health node", currentHealth);
```
Done.

gmarouli commented on server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java:
```diff
@@ -0,0 +1,233 @@
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        // Wait until every node in the cluster is upgraded to 8.4.0 or later
```
Ah, I thought it was already covered by the check for the version. I will take a look and if it's not then we should change it in all the other places we have been checking the version. Right?

andreidan replied:
Ah righto - `nodesIfRecovered` does this check already TIL

gmarouli commented on server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java:
```diff
@@ -0,0 +1,233 @@
+        if (inProgress.compareAndSet(false, true)) {
+            ClusterState clusterState = clusterService.state();
+            HealthMetadata healthMetadata = HealthMetadata.getFromClusterState(clusterState);
+            assert healthMetadata != null : "health metadata should have been initialized.";
```
Before we schedule any monitoring task we check the flag `healthMetadataInitialized` so theoretically this should hold true. The only case I can think of, is if someone removes the health metadata. But I think we do not give that option programmatically. Would you prefer to have an `if` here?

andreidan replied:
Ah, right. The `HealthMetadataService` is controlled by a different setting - enabled/disabled via `health.node.enabled` so that made me think things can diverge. 
I prefer the explicit checks, but it's a personal preference so feel free to ignore

gmarouli replied:
They are all controlled by the same flag `health.node.enabled`. The reason I have done this, is that if the health node is disabled, there is no point to monitor the disk health locally, or to publish health metadata right?

On a second thought, depending on the use case @jbaiera is working on we might want to decouple this and have the health metadata always be published. They might be useful for the coordinator node too.

gmarouli committed:
Merge branch 'main' into health-disk-monitor-health

gmarouli committed:
Polishing after review

gmarouli committed:
Remove IndividualNodeHealth.java

gmarouli committed:
Merge branch 'main' into health-disk-monitor-health

gmarouli committed:
Document synchronizing with inProgress flag.

gmarouli committed:
Fix version

gmarouli committed:
Rename healthMetadataInitialized to be inclusive

andreidan reviewed (commented):
Thanks for iterating on this. 

Left a could more comments

gmarouli commented on server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java:
```diff
@@ -0,0 +1,201 @@
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (prerequisitesFulfilled == false) {
+            prerequisitesFulfilled = event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_5_0)
+                && HealthMetadata.getFromClusterState(event.state()) != null;
+            scheduleNowIfEnabled();
+        }
```
Hm, I think I need to rename this method. We check the prerequisites in the method `scheduleNowIfEnabled`. This was a choice I made to ensure the prerequisites are always checked before any scheduling. For this reason it felt redundant to check it again here. What do you think about renaming to `maybeScheduleNow()` indicating that it is not a given that it will schedule it?

gmarouli replied:
About `8.5.0` you are right the `healthMetadata` do exists in `8.4.0` but this local monitor will be pushing changes to the `HealthNode` and that endpoint is not implemented in `8.4.0`. That's why I thought that the minimum required version should be `8.5.0`. Does this make sense?

andreidan replied:
Ah ++ fair enough

gmarouli commented on server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java:
```diff
@@ -0,0 +1,201 @@
+    // monitoring tasks scheduled, one of them will be no-op.
+    private final AtomicBoolean inProgress = new AtomicBoolean();
+    // Keeps the latest health state that was successfully reported.
+    private DiskHealthInfo lastReportedDiskHealthInfo = null;
```
Yeap, I probably removed it accidentally at some point.

gmarouli committed:
Clarify scheduling, make lastReported volatile

andreidan reviewed (approved):
LGTM thanks for iterating on this Mary 🚀

gmarouli commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into health-disk-monitor-health

gmarouli commented:
@elasticmachine run elasticsearch-ci/part-1

gmarouli commented:
@andreidan thanks for the feedback!
